Dev/issue 8635 fixity logging

mcantelon opened the pull request:
This adds:
- failure handling if the fixity check doesn't complete because of an exception
- per-AIP log model of fixity attempts (time, whether they're successful, and error/failure details)
- two columns to package model for latest fixity attempt datetime/result
- two columns in package hitlist for latest fixity attempt datetime/result
- a fixity failures page where fixity check history for a package can be viewed

I'm not sure how you folk merge (whether you squash or merge individual commits), but I can clean up the history and put in better commit messages if individual commits are merged.

Hwesta commented on storage_service/locations/api/resources.py:
```diff
@@ -670,7 +671,12 @@ def pointer_file_request(self, request, bundle, **kwargs):
-        success, failures, message = bundle.obj.check_fixity()
+        try:
+            success, failures, message = bundle.obj.check_fixity()
+        except:
```
Only catch the exception(s) you're expecting here, or at a minimum `except Exception`.  Maybe a logging message with a traceback too?

Hwesta commented on storage_service/locations/models/fixity_log.py:
```diff
@@ -0,0 +1,28 @@
+    package = models.ForeignKey('Package', to_field='uuid')
+    success = models.BooleanField(default=False)
+    error_details = models.TextField()
+    time_reported = models.DateTimeField(auto_now=True)
```
Nitpick, but the `DateTimeField` is called `time` here and `date` in Package. For consistency, call them both the same thing. Maybe datetime?

Hwesta commented on storage_service/locations/signals.py:
```diff
@@ -44,4 +47,26 @@ def report_failed_fixity_check(sender, **kwargs):
+        package.last_fixity_check_date = log.time_reported
+        package.save()
+    except:
+        LOGGER.exception("Error logging fixity failure.")
```
Is this an ERROR level log or something lesser?  Tracebacks can be added to any log message with [exc_info](https://docs.python.org/2/library/logging.html#logging.debug)

Hwesta commented on storage_service/locations/signals.py:
```diff
@@ -44,4 +47,26 @@ def report_failed_fixity_check(sender, **kwargs):
+    package = Package.objects.get(uuid=bundle.obj.uuid)
+    package.last_fixity_check_result = True
+    package.last_fixity_check_date = datetime.datetime.now()
+    package.save()
```
Should a FixityLog be created here as well?

Hwesta commented on storage_service/locations/views.py:
```diff
@@ -46,6 +46,11 @@ def package_list(request):
 
+def package_fixity(request, uuid):
+    package = Package.objects.get(uuid=uuid)
+    log_entries = FixityLog.objects.filter(package=package)
```
This is easier with `FixityLog.objects.filter(package__uuid=uuid)` unless you need the package as well.  See https://docs.djangoproject.com/en/dev/topics/db/queries/#lookups-that-span-relationships for details.  Objects also store [reverse-foreign-key-sets](https://docs.djangoproject.com/en/dev/topics/db/queries/#following-relationships-backward).

Hwesta commented on storage_service/locations/views.py:
```diff
@@ -46,6 +46,11 @@ def package_list(request):
     packages = Package.objects.all()
     return render(request, 'locations/package_list.html', locals())
 
+def package_fixity(request, uuid):
```
Be wary of shadowing common libraries like `uuid`, since it can cause weird bugs later.

Hwesta commented on storage_service/locations/views.py:
```diff
@@ -46,6 +46,11 @@ def package_list(request):
+def package_fixity(request, uuid):
+    package = Package.objects.get(uuid=uuid)
+    log_entries = FixityLog.objects.filter(package=package)
+    return render(request, 'locations/fixity_results.html', locals())
```
I know it's super convenient, but we're trying to move away from using `locals()` and towards an explicit dict of the variables used.

Hwesta commented on storage_service/locations/models/package.py:
```diff
@@ -91,6 +91,10 @@ class Package(models.Model):
 
+    # Provide access, without join, to last fixity check result/date
+    last_fixity_check_result = models.NullBooleanField()
+    last_fixity_check_date = models.DateTimeField(blank=True, null=True)
```
Are these duplicates of `FixityLog.[success|time_reported]`?  I'm concerned about them getting out of sync.  Maybe make it a foreign key and look at this performance tweak if it becomes a problem?  Are there case where we need this info but no FixityLog object exists?

Hwesta commented:
> I'm not sure how you folk merge (whether you squash or merge individual commits), but I can clean up the history and put in better commit messages if individual commits are merged.

Generally we do a rebase on top of qa/0.x, then merge it as a fast-forward merge. A cleaner commit history would be appreciated.  https://github.com/artefactual/archivematica/blob/code_review_guide/code_review.md maybe useful as well.

Hwesta commented on storage_service/locations/signals.py:
```diff
@@ -1,8 +1,11 @@
+
 LOGGER = logging.getLogger(__name__)
 
 deletion_request = Signal(providing_args=["uuid", "location", "url"])
```
`successful_fixity_check` needs to be added here.

Hwesta commented:
I strongly recommend installing a linter (probably [flake8](https://flake8.readthedocs.org/en/2.4.1/)) in your text editor, since it caught the undefined variable above.

Also, we have tests now!  Run `py.test --ignore locations/tests/test_models.py` from the `storage_service` directory

mistydemeo commented on storage_service/locations/api/resources.py:
```diff
@@ -670,7 +671,12 @@ def pointer_file_request(self, request, bundle, **kwargs):
+        except:
+            success = False
+            failures = []
+            message = 'Internal error'
```
Probably good to a) log the error, b) include some additional information in the response. We don't need to expose too much internal app state in the API response, but it'd be good to get an idea _why_ it didn't work.

mistydemeo commented on storage_service/locations/signals.py:
```diff
@@ -44,4 +47,26 @@ def report_failed_fixity_check(sender, **kwargs):
+        package.last_fixity_check_result = False
+        package.last_fixity_check_date = log.time_reported
+        package.save()
+    except:
```
Like before, are there specific exceptions to catch here?

mcantelon committed:
Add fixity log entry model.

This model will be used to support logging of fixity check results.

mcantelon committed:
Log fixity check results

Where endpoint for fixity checks is requested, log the results of the check.

mcantelon committed:
Added fixity status/log display

Added display of the last fixity check date/result in the package list page.
Each result links to a page showing a log of fixity checks on the package.

Hwesta commented on storage_service/locations/signals.py:
```diff
@@ -44,4 +48,23 @@ def report_failed_fixity_check(sender, **kwargs):
+
+    log = models.FixityLog()
+    log.package = package
+    report_data = json.loads(kwargs['report'])
```
Is this supposed to be used?

Hwesta commented on storage_service/locations/signals.py:
```diff
@@ -44,4 +48,23 @@ def report_failed_fixity_check(sender, **kwargs):
+    log.package = package
+    report_data = json.loads(kwargs['report'])
+    log.error_details = report_data['message']
+    log.save()
```
You can also create an object in one line/atomically with [`FixityLog.objects.create()`](https://docs.djangoproject.com/en/dev/ref/models/querysets/#create)

Hwesta commented on storage_service/locations/signals.py:
```diff
@@ -44,4 +48,23 @@ def report_failed_fixity_check(sender, **kwargs):
+    log.package = package
+    report_data = json.loads(kwargs['report'])
+    log.success = True
+    log.save()
```
Could this be merged with `report_failed_fixity_check` above? They seem very similar.  Also, report_data is not used in this function.

Hwesta commented on storage_service/locations/views.py:
```diff
@@ -44,8 +44,23 @@ def get_delete_context_dict(request, model, object_uuid, default_cancel='/'):
+        package_data.append((latest_check, latest_result, package))
+
+    context = {'package_data': package_data}
     return render(request, 'locations/package_list.html', locals())
```
You probably meant `return render(request, 'locations/package_list.html', {'package_data': package_data, 'package': package)`

Hwesta commented on storage_service/templates/locations/fixity_results.html:
```diff
@@ -0,0 +1,28 @@
+    </tbody>
+  </table>
+{% else %}
+  <p>No fixity failures exist.</p>
```
This isn't just for errors anymore, is it?  Perhaps display the `entry.success` value in the table too.

Hwesta commented on storage_service/locations/views.py:
```diff
@@ -44,8 +44,23 @@ def get_delete_context_dict(request, model, object_uuid, default_cancel='/'):
+        fixity_checks = FixityLog.objects.filter(package=package).order_by('-datetime_reported')
+        latest_check = fixity_checks[0].datetime_reported if fixity_checks else None
+        latest_result = fixity_checks[0].success if fixity_checks else None
+        package_data.append((latest_check, latest_result, package))
```
Since we need this info in several places (the package list, the Locations list where it shows several packages) then we probably want to generate this information in once place.  I think it makes sense to add latest_check & latest_result to the Package as @propeties.  Eg.

``` python
class Package(models.Model):
    ...
    @property
    def latest_fixity_check(self):
        try:
            latest_check = FixityLog.objects.filter(package=self).order_by('-datetime_reported')[0]  # limit 1
        except IndexError:
            return None
        return latest_check.datetime_reported
```

Thoughts?

Hwesta replied:
I see you added the fixity attributes to the `Package` class, this view should be updated to use them instead of re-generating the information.  The template also needs to be updated for the new attributes.  `render` should return the `context` that it uses, not `locals()`.

Hwesta commented:
I noticed several try/except blocks disappeared in the rebase - why?

Also, tests?

Hwesta commented on storage_service/locations/signals.py:
```diff
@@ -35,8 +39,16 @@ def report_deletion_request(sender, **kwargs):
 
+def _log_report(uuid, success, message=None):
+    package = models.Package.objects.get(uuid=uuid)
+    models.FixityLog.objects.create(package=package, error_details=message)
```
Shouldn't `success` be used here?

This can also be done with `models.FixityLog.objects.create(package_id=uuid, error_details=message)`.  Is it worth having a 1-line cleanup function?

Hwesta commented on storage_service/templates/snippets/packages_table.html:
```diff
@@ -9,11 +9,13 @@
     </thead>
     <tbody>
-    {% for package in packages %}
+    {% for last_fixity_check_datetime, last_fixity_check_result, package in package_data %}
```
`last_fixity_check_datetime, last_fixity_check_result` are never provided as context anywhere - is this leftover from previous changes?

Hwesta commented on storage_service/locations/models/package.py:
```diff
@@ -163,6 +163,28 @@ def is_compressed(self):
+
+    def _latest_fixity_check(self):
+        try:
+            return FixityLog.objects.filter(package=self).order_by('-datetime_reported')[0]  # limit 1
```
Please run a linter over your code before you commit it. It will find things like `FixityLog` being undefined.

mcantelon committed:
Added fixity log tests.

Hwesta commented on storage_service/locations/views.py:
```diff
@@ -43,8 +43,13 @@ def get_delete_context_dict(request, model, object_uuid, default_cancel='/'):
+    return render(request, 'locations/package_list.html', context)
+
+def package_fixity(request, package_uuid):
+    log_entries = FixityLog.objects.filter(package__uuid=package_uuid).order_by('-datetime_reported')
```
Tip: You can filter by the value in a foreign key (instead of the object, which requires a join) by using `_id` https://docs.djangoproject.com/en/1.8/topics/db/queries/#field-lookups  So this could be `FixityLog.objects.filter(package_id=package_uuid).order_by('-datetime_reported')`

Hwesta commented on storage_service/locations/tests/test_fixity_log.py:
```diff
@@ -0,0 +1,18 @@
+
+class TestFixityLog(TestCase):
+
+    fixtures = ['base.json', 'fixity_log.json']
```
This fixture fails to load with `IntegrityError: Problem installing fixtures: The row in table 'locations_fixitylog' with primary key '1' has an invalid foreign key: locations_fixitylog.package_id contains a value 'e0a41934-c1d7-45ba-9a95-a7531c063ed1' that does not have a corresponding value in locations_package.uuid.`  You need to add the 'package.json' to the fixtures list.

Hwesta commented on storage_service/locations/tests/test_fixity_log.py:
```diff
@@ -0,0 +1,18 @@
+
+    def setUp(self):
+        self.fl_object = models.FixityLog.objects.all()[0]
+        #self.auth = requests.auth.HTTPBasicAuth(self.ds_object.user, self.ds_object.password)
```
This looks leftover?

Hwesta commented:
Glad to see tests! Maybe add a couple for the Package's new attributes (since they're actually functions) `latest_fixity_check_datetime` `latest_fixity_check_result`, especially if there's no FixityLog objects.

Also, please rebase this on qa/0.x

mcantelon commented:
I got rid of some of the try/except blocks because I couldn't remember what had caused them and hence couldn't be specific about what exception to check for.

sevein commented:
@mcantelon, it looks like it was a mistake to add this to the `0.8.0` milestone. Sorry for the trouble.

mcantelon commented:
Merged in qa/0.x.
